agent/src/agent/pipeline/manager.py

```diff
@@ -313,14 +323,30 @@ def transform_for_bc(pipeline_: Pipeline) -> dict:
 
 
 def should_send_error_notification(pipeline_: Pipeline) -> bool:
-    # number of error statuses = number of retries + 1
     return not constants.DISABLE_PIPELINE_ERROR_NOTIFICATIONS \
-           and pipeline_.error_notification_enabled() \
-           and pipeline_.retries \
+           and pipeline_.error_notification_enabled()
+
+
+def should_send_retries_error_notification(pipeline_: Pipeline) -> bool:
+    # number of error statuses = number of retries + 1
+    return should_send_error_notification(pipeline_) \
+           and bool(pipeline_.retries) \
            and pipeline_.retries.number_of_error_statuses - 1 >= constants.STREAMSETS_NOTIFY_AFTER_RETRY_ATTEMPTS \
            and not pipeline_.retries.notification_sent
 
 
+def should_send_no_data_error_notification(pipeline_: Pipeline) -> bool:
+    if pipeline_.notifications and pipeline_.notifications.no_data_notification:
+        no_data_notification_period = timedelta(
+            minutes=pipeline_.notifications.no_data_notification.notification_period
+        ) + timedelta(seconds=pipeline_.interval or 0)
+        return should_send_error_notification(pipeline_) \
+               and not pipeline_.notifications.no_data_notification.notification_sent \
+               and bool(pipeline_.watermark) \
```

Comment:
this should be compared by last sent offset (sometimes we sent watermark even if there is no data, for example, when dvp is enabled)